File: src/freight_fate/states/driving_stop_detail.py
from __future__ import annotations

from typing import TYPE_CHECKING

from ..sim.trip_models import RoadStop
from .base import MenuItem, MenuState
from .driving_core import POI_ACTION_LABELS, POI_SERVICE_LABELS, _join_phrase

if TYPE_CHECKING:
    from .driving import DrivingState

# Matches Trip.eta_game_hours: parked or crawling assumes a highway pace.
FALLBACK_MPH = 55.0


class StopDetailState(MenuState):
    """Full details for one upcoming route stop, opened from the Map screen.

    Mirrors the dispatch board's job details view: each fact is its own
    menu row, then the plan/cancel buttons, then Back.
    """

    title = "Stop details"
    intro_help = (
        "Use up and down arrows to review each detail line; Home and End "
        "jump to the first and last row. Enter repeats detail lines or "
        "activates the planning buttons. Escape returns to the map screen."
    )

    def __init__(self, ctx, driving: DrivingState, stop: RoadStop) -> None:
        super().__init__(ctx)
        self.driving = driving
        self.stop = stop

    def enter(self) -> None:
        self.items = self.build_items()
        self.index = min(self.index, max(0, len(self.items) - 1))
        self.ctx.audio.play(self.open_sound_key)
        self.announce_entry()

    def announce_entry(self) -> None:
        self.ctx.say(f"Stop details. {self.intro_help} {self.current_text()}")

    def current_help(self) -> str:
        return f"{self.intro_help} {super().current_help()}"

    def build_items(self) -> list[MenuItem]:
        trip = self.driving.trip
        items = [
            MenuItem(
                line,
                lambda line=line: self.ctx.say(line),
                help="This is a stop detail line. Press Enter to repeat it.",
            )
            for line in self._detail_lines()
        ]
        planned = trip.planned_stop_key
        if trip.is_planned(self.stop):
            items.append(
                MenuItem(
                    f"Cancel planned stop at {self.stop.name}",
                    self._cancel,
                    help="Forget this planned stop. Announcements go back to normal.",
                )
            )
        else:
            move_note = (
                " You already have a planned stop; you'll be asked to confirm moving it here."
                if planned is not None
                else ""
            )
            items.append(
                MenuItem(
                    f"Plan to stop at {self.stop.name}",
                    self._plan,
                    help="Remember this stop. Announcements will call it your "
                    f"planned stop so you know when to take the exit.{move_note}",
                    # _plan (or the move confirmation) plays its own chime.
                    select_sound=None,
                )
            )
        items.append(MenuItem("Back", self.go_back, help="Return to the map screen."))
        return items

    def _detail_lines(self) -> list[str]:
        stop = self.stop
        trip = self.driving.trip
        settings = self.ctx.settings
        ahead = max(0.0, stop.at_mi - trip.position_mi)
        lines = [f"Stop: {trip.planned_prefix(stop)}{stop.spoken_name}."]
        if stop.exit_label:
            lines.append(f"Exit: {stop.exit_label}.")
        lines.append(f"Distance: {settings.distance_text(ahead)} ahead.")
        offers = [
            POI_ACTION_LABELS[action] for action in stop.actions if action in POI_ACTION_LABELS
        ]
        services = [
            POI_SERVICE_LABELS.get(service, service.replace("_", " ")) for service in stop.services
        ]
        if offers:
            lines.append(f"Offers: {_join_phrase(offers)}.")
        if services:
            lines.append(f"Listed services: {_join_phrase(services)}.")
        if not offers and not services:
            lines.append("Services not listed.")
        if stop.parking_text:
            lines.append(f"Parking: {stop.parking_text}.")
        lines.append(self._eta_line(ahead))
        return lines

    def _eta_line(self, ahead: float) -> str:
        # Same rule as Trip.eta_game_hours, over the distance to this stop.
        speed = self.driving.truck.speed_mph
        trip = self.driving.trip
        if speed >= trip.ETA_MIN_MPH:
            mph = speed
            basis = "at your current speed"
        else:
            mph = FALLBACK_MPH
            basis = "at a typical highway pace"
        eta_h = ahead / max(1.0, mph)
        hos_note = self.driving.hos.arrival_note(self.ctx.settings.hos_mode, eta_h * 60.0)
        return f"Estimated time to reach it: {eta_h:.1f} hours {basis}.{hos_note}"

    def _plan(self) -> None:
        trip = self.driving.trip
        if trip.planned_stop_key is not None and not trip.is_planned(self.stop):
            # A different stop is already planned: confirm before moving it.
            self.ctx.push_state(ConfirmMovePlanState(self.ctx, self.driving, self.stop))
            return
        self._set_planned_stop()

    def _set_planned_stop(self) -> None:
        self.driving.trip.planned_stop_key = self.stop.key
        self.ctx.audio.play("ui/notify")
        self.refresh()
        self.ctx.say(
            f"Planned stop set: {self.stop.spoken_name}. You will hear it "
            "called your planned stop as you approach."
        )

    def _cancel(self) -> None:
        self.driving.trip.planned_stop_key = None
        self.refresh()
        self.ctx.say("Planned stop canceled.")


class ConfirmMovePlanState(MenuState):
    """Yes/No guard shown when planning a stop while another stop is planned.

    Names the current planned stop and how far ahead it is, then asks whether
    to move the plan here. Lands on "Yes" so one Enter completes the move the
    player just asked for.
    """

    title = "Move planned stop?"
    intro_help = (
        "Use up and down arrows to navigate, Enter to select. "
        "Escape keeps your current planned stop."
    )

    def __init__(self, ctx, driving: DrivingState, stop: RoadStop) -> None:
        super().__init__(ctx)
        self.driving = driving
        self.stop = stop

    def _planned_stop(self) -> RoadStop | None:
        return self.driving.trip.planned_stop

    def _ahead_text(self, stop: RoadStop) -> str:
        ahead = max(0.0, stop.at_mi - self.driving.trip.position_mi)
        return self.ctx.settings.distance_text(ahead)

    def announce_entry(self) -> None:
        current = self._planned_stop()
        where = (
            f"{current.spoken_name}, {self._ahead_text(current)} ahead"
            if current is not None
            else self.driving.trip.planned_stop_label
        )
        self.ctx.say(
            f"{self.title} You already have a planned stop at {where}. "
            f"Move your plan to {self.stop.spoken_name}, "
            f"{self._ahead_text(self.stop)} ahead? {self.current_text()}"
        )

    def build_items(self) -> list[MenuItem]:
        current = self._planned_stop()
        keep_name = (
            current.spoken_name if current is not None else self.driving.trip.planned_stop_label
        )
        return [
            MenuItem(
                f"Yes, move plan to {self.stop.spoken_name}",
                self._confirm,
                help="Move your planned stop to this one.",
            ),
            MenuItem(
                f"No, keep planned stop at {keep_name}",
                self.go_back,
                help="Return to the stop details without moving your plan.",
            ),
        ]

    def _confirm(self) -> None:
        self.driving.trip.planned_stop_key = self.stop.key
        # Popping re-enters the detail screen, which now shows the Cancel button.
        self.ctx.pop_state()
        self.ctx.audio.play("ui/notify")
        self.ctx.say(
            f"Planned stop moved to {self.stop.spoken_name}. You will hear it "
            "called your planned stop as you approach.",
            interrupt=True,  # cut the detail screen's re-entry announcement
        )

File: crates/freight-fate/src/states/driving_stop_detail.rs
//! Full details for one upcoming route stop, and the guard in front of
//! moving a planned stop (port of
//! `freight_fate/states/driving_stop_detail.py`).

use ff_core::pyfmt::fmt_f;
use ff_core::sim::trip::ETA_MIN_MPH;
use ff_core::sim::trip_models::RoadStop;

use crate::app::{GameContext, Say};
use crate::impl_state_for_menu;
use crate::states::base::{Menu, MenuCore, MenuItem};
use crate::states::driving_core::{hos_of, join_phrase, POI_ACTION_LABELS, POI_SERVICE_LABELS};
use crate::states::driving_menu_states::DriveRef;

/// Matches `Trip.eta_game_hours`: parked or crawling assumes a highway pace.
pub const FALLBACK_MPH: f64 = 55.0;

const STOP_DETAIL_INTRO_HELP: &str =
    "Up and down arrows review each line; Home and End jump to the first and last. Enter repeats \
     a line or activates a button. Escape returns to the map.";

/// Full details for one upcoming route stop, opened from the Map screen.
///
/// Mirrors the dispatch board's job details view: each fact is its own menu
/// row, then the plan/cancel buttons, then Back.
pub struct StopDetailState {
    menu: MenuCore<Self>,
    driving: DriveRef,
    pub stop: RoadStop,
}

impl StopDetailState {
    pub fn new(driving: DriveRef, stop: RoadStop) -> Self {
        StopDetailState {
            menu: MenuCore::new("Stop details").with_intro_help(STOP_DETAIL_INTRO_HELP),
            driving,
            stop,
        }
    }

    fn detail_lines(&mut self, ctx: &mut GameContext) -> Vec<String> {
        let stop = self.stop.clone();
        self.driving
            .with(ctx, |d, ctx| {
                let ahead = (stop.at_mi - d.trip.position_mi).max(0.0);
                let mut lines = vec![format!(
                    "Stop: {}{}.",
                    d.trip.planned_prefix(&stop),
                    stop.spoken_name()
                )];
                if !stop.exit_label.is_empty() {
                    lines.push(format!("Exit: {}.", stop.exit_label));
                }
                lines.push(format!(
                    "Distance: {} ahead.",
                    ctx.settings.distance_text(ahead, false)
                ));
                let offers: Vec<String> = stop
                    .actions
                    .iter()
                    .filter_map(|action| {
                        POI_ACTION_LABELS
                            .iter()
                            .find(|(key, _)| key == action)
                            .map(|(_, label)| label.to_string())
                    })
                    .collect();
                let services: Vec<String> = stop
                    .services
                    .iter()
                    .map(|service| {
                        POI_SERVICE_LABELS
                            .iter()
                            .find(|(key, _)| key == service)
                            .map(|(_, label)| label.to_string())
                            .unwrap_or_else(|| service.replace('_', " "))
                    })
                    .collect();
                if !offers.is_empty() {
                    lines.push(format!("Offers: {}.", join_phrase(&offers)));
                }
                if !services.is_empty() {
                    lines.push(format!("Listed services: {}.", join_phrase(&services)));
                }
                if offers.is_empty() && services.is_empty() {
                    lines.push("Services not listed.".to_string());
                }
                let parking_text = stop.parking_text();
                if !parking_text.is_empty() {
                    lines.push(format!("Parking: {parking_text}."));
                }
                lines.push(eta_line(d, ctx, ahead));
                lines
            })
            .unwrap_or_default()
    }

    fn plan(&mut self, ctx: &mut GameContext) {
        let stop = self.stop.clone();
        let needs_confirm = self
            .driving
            .read(|d| d.trip.planned_stop_key.is_some() && !d.trip.is_planned(&stop))
            .unwrap_or(false);
        if needs_confirm {
            // A different stop is already planned: confirm before moving it.
            let state = ConfirmMovePlanState::new(self.driving.clone(), self.stop.clone());
            ctx.push_state(state);
            return;
        }
        self.set_planned_stop(ctx);
    }

    fn set_planned_stop(&mut self, ctx: &mut GameContext) {
        let key = self.stop.key();
        self.driving.read(|d| d.trip.planned_stop_key = Some(key));
        ctx.audio.play("ui/notify");
        self.refresh(ctx, true);
        let name = self.stop.spoken_name();
        ctx.say(&format!("Planned stop set, {name}."));
    }

    fn cancel(&mut self, ctx: &mut GameContext) {
        self.driving.read(|d| d.trip.planned_stop_key = None);
        self.refresh(ctx, true);
        ctx.say("Planned stop canceled.");
    }
}

/// Same rule as `Trip.eta_game_hours`, over the distance to this stop.
fn eta_line(d: &mut crate::states::driving::DrivingState, ctx: &GameContext, ahead: f64) -> String {
    let speed = d.trip.truck.speed_mph();
    let (mph, basis) = if speed >= ETA_MIN_MPH {
        (speed, "at your current speed")
    } else {
        (FALLBACK_MPH, "at a typical highway pace")
    };
    let eta_h = ahead / mph.max(1.0);
    let hos_note = hos_of(ctx).arrival_note(&ctx.settings.hos_mode, eta_h * 60.0);
    format!(
        "Estimated time to reach it: {} hours {basis}.{hos_note}",
        fmt_f(eta_h, 1)
    )
}

impl Menu for StopDetailState {
    fn menu(&self) -> &MenuCore<Self> {
        &self.menu
    }

    fn menu_mut(&mut self) -> &mut MenuCore<Self> {
        &mut self.menu
    }

    fn build_items(&mut self, ctx: &mut GameContext) -> Vec<MenuItem<Self>> {
        let mut items: Vec<MenuItem<Self>> = self
            .detail_lines(ctx)
            .into_iter()
            .map(|line| {
                let spoken = line.clone();
                MenuItem::new(line, move |_s: &mut Self, ctx: &mut GameContext| {
                    ctx.say(&spoken)
                })
                .help("Enter repeats this line.")
            })
            .collect();
        let stop = self.stop.clone();
        let planned = self
            .driving
            .read(|d| d.trip.planned_stop_key.clone())
            .flatten();
        let is_planned = self
            .driving
            .read(|d| d.trip.is_planned(&stop))
            .unwrap_or(false);
        if is_planned {
            items.push(
                MenuItem::new(
                    format!("Cancel planned stop at {}", self.stop.name),
                    |s: &mut Self, ctx| s.cancel(ctx),
                )
                .help("Forgets this planned stop."),
            );
        } else {
            let move_note = if planned.is_some() {
                " A planned stop already exists; you confirm moving it here."
            } else {
                ""
            };
            items.push(
                MenuItem::new(
                    format!("Plan to stop at {}", self.stop.name),
                    |s: &mut Self, ctx| s.plan(ctx),
                )
                .help(format!(
                    "Plans this stop; announcements call it your planned stop.{move_note}"
                ))
                // `plan` (or the move confirmation) plays its own chime.
                .select_sound(None),
            );
        }
        items.push(
            MenuItem::new("Back", |s: &mut Self, ctx| s.go_back(ctx))
                .help("Return to the map screen."),
        );
        items
    }

    fn announce_entry(&mut self, ctx: &mut GameContext) {
        let current = self.current_text(ctx);
        ctx.say(&format!("Stop details. {STOP_DETAIL_INTRO_HELP} {current}"));
    }

    fn current_help(&self, ctx: &GameContext) -> String {
        let core = self.menu();
        let base = if core.items.is_empty() {
            core.intro_help.clone()
        } else {
            let item = &core.items[core.index];
            let help = item.help_text(self, ctx);
            if help.is_empty() {
                format!("{}.", item.text(self, ctx))
            } else {
                help
            }
        };
        format!("{STOP_DETAIL_INTRO_HELP} {base}")
    }
}

impl_state_for_menu!(StopDetailState);

/// Yes/No guard shown when planning a stop while another stop is planned.
///
/// Names the current planned stop and how far ahead it is, then asks whether
/// to move the plan here. Lands on "Yes" so one Enter completes the move the
/// player just asked for.
pub struct ConfirmMovePlanState {
    menu: MenuCore<Self>,
    driving: DriveRef,
    pub stop: RoadStop,
}

const MOVE_PLAN_INTRO_HELP: &str = "Use up and down arrows to navigate, Enter to select. \
                                    Escape keeps your current planned stop.";

impl ConfirmMovePlanState {
    pub fn new(driving: DriveRef, stop: RoadStop) -> Self {
        ConfirmMovePlanState {
            menu: MenuCore::new("Move planned stop?").with_intro_help(MOVE_PLAN_INTRO_HELP),
            driving,
            stop,
        }
    }

    fn planned_stop(&self) -> Option<RoadStop> {
        self.driving
            .read(|d| d.trip.planned_stop().cloned())
            .flatten()
    }

    fn ahead_text(&self, ctx: &GameContext, stop: &RoadStop) -> String {
        let ahead = self
            .driving
            .read(|d| (stop.at_mi - d.trip.position_mi).max(0.0))
            .unwrap_or(0.0);
        ctx.settings.distance_text(ahead, false)
    }

    fn confirm(&mut self, ctx: &mut GameContext) {
        let key = self.stop.key();
        self.driving.read(|d| d.trip.planned_stop_key = Some(key));
        // Popping re-enters the detail screen, which now shows the Cancel
        // button.
        ctx.pop_state();
        ctx.audio.play("ui/notify");
        let name = self.stop.spoken_name();
        ctx.say_with(
            format!("Planned stop moved to {name}."),
            // interrupt: cut the detail screen's re-entry announcement
            Say::new(),
        );
    }
}

impl Menu for ConfirmMovePlanState {
    fn menu(&self) -> &MenuCore<Self> {
        &self.menu
    }

    fn menu_mut(&mut self) -> &mut MenuCore<Self> {
        &mut self.menu
    }

    fn build_items(&mut self, _ctx: &mut GameContext) -> Vec<MenuItem<Self>> {
        let keep_name = match self.planned_stop() {
            Some(stop) => stop.spoken_name(),
            None => self
                .driving
                .read(|d| d.trip.planned_stop_label())
                .unwrap_or_default(),
        };
        vec![
            MenuItem::new(
                format!("Yes, move plan to {}", self.stop.spoken_name()),
                |s: &mut Self, ctx| s.confirm(ctx),
            )
            .help("Move your planned stop to this one."),
            MenuItem::new(
                format!("No, keep planned stop at {keep_name}"),
                |s: &mut Self, ctx| s.go_back(ctx),
            )
            .help("Return to the stop details without moving your plan."),
        ]
    }

    fn announce_entry(&mut self, ctx: &mut GameContext) {
        let where_ = match self.planned_stop() {
            Some(current) => format!(
                "{}, {} ahead",
                current.spoken_name(),
                self.ahead_text(ctx, &current)
            ),
            None => self
                .driving
                .read(|d| d.trip.planned_stop_label())
                .unwrap_or_default(),
        };
        let title = self.menu.title.clone();
        let target = self.stop.spoken_name();
        let stop = self.stop.clone();
        let ahead = self.ahead_text(ctx, &stop);
        let current = self.current_text(ctx);
        ctx.say(&format!(
            "{title} You already have a planned stop at {where_}. Move your plan to {target}, \
             {ahead} ahead? {current}"
        ));
    }
}

impl_state_for_menu!(ConfirmMovePlanState);
